sgm/inference/helpers.py

```diff
@@ -276,30 +314,74 @@ def do_img2img(
+
                 sigmas = sampler.discretization(sampler.num_steps)
                 sigma = sigmas[0].to(z.device)
 
                 if offset_noise_level > 0.0:
                     noise = noise + offset_noise_level * append_dims(
                         torch.randn(z.shape[0], device=z.device), z.ndim
                     )
-                noised_z = z + noise * append_dims(sigma, z.ndim)
-                noised_z = noised_z / torch.sqrt(
-                    1.0 + sigmas[0] ** 2.0
-                )  # Note: hardcoded to DDPM-like scaling. need to generalize later.
+                if add_noise:
+                    noised_z = z + noise * append_dims(sigma, z.ndim).cuda()
+                    noised_z = noised_z / torch.sqrt(
+                        1.0 + sigmas[0] ** 2.0
+                    )  # Note: hardcoded to DDPM-like scaling. need to generalize later.
+                else:
+                    noised_z = z / torch.sqrt(1.0 + sigmas[0] ** 2.0)
 
                 def denoiser(x, sigma, c):
                     return model.denoiser(model.model, x, sigma, c)
 
-                samples_z = sampler(denoiser, noised_z, cond=c, uc=uc)
-                samples_x = model.decode_first_stage(samples_z)
-                samples = torch.clamp((samples_x + 1.0) / 2.0, min=0.0, max=1.0)
+                with SwapToDevice(model.denoiser, device):
+                    with SwapToDevice(model.model, device):
+                        samples_z = sampler(denoiser, noised_z, cond=c, uc=uc)
+
+                with SwapToDevice(model.first_stage_model, device):
+                    samples_x = model.decode_first_stage(samples_z)
+                    samples = torch.clamp((samples_x + 1.0) / 2.0, min=0.0, max=1.0)
 
                 if filter is not None:
                     samples = filter(samples)
 
                 if return_latents:
                     return samples, samples_z
                 return samples
+
+
+@contextlib.contextmanager
+def SwapToDevice(
+    model: Union[torch.nn.Module, torch.Tensor], device: Union[torch.device, str]
+):
+    """
+    Context manager that swaps a model or tensor to a device, and then swaps it back to its original device
+    when the context is exited.
+    """
+    if isinstance(model, torch.Tensor):
+        original_device = model.device
+    else:
+        param = next(model.parameters(), None)
+        if param is not None:
+            original_device = param.device
+        else:
+            buf = next(model.buffers(), None)
+            if buf is not None:
+                original_device = buf.device
+            else:
+                # If device could not be found, do nothing
+                return
+    device = torch.device(device)
+
+    if device != original_device:
+        model.to(device)
+
+    yield
+
+    if device != original_device:
+        model.to(original_device)
+        if torch.cuda.is_available():
+            torch.cuda.empty_cache()
```

Comment:
```suggestion
    try:
        if device != original_device:
            model.to(device)

        yield
    finally:
        if device != original_device:
            model.to(original_device)
            if torch.cuda.is_available():
                torch.cuda.empty_cache()
```
so things are cleaned up if the block this is used in fails.